csrc/scheduler/hopper_multi_matmul.cpp

```diff
@@ -507,40 +507,49 @@ void HopperMultipleMatmulScheduler::scheduleEpilogue() {
       // NOTE: cacheBefore does not work with blockTileTensors
       TensorView* d_smem = cacheAfter(dc, LoadStoreOpType::Set);
 
-      std::vector<TensorView*> tvs_to_schedule{d, d_smem};
-      if (std::find(mma_results_.begin(), mma_results_.end(), dc) ==
-          mma_results_.end()) {
-        // Skip scheduling dc if it is an mma_result. This can happen if we are
-        // not casting back to half-precision in the output
-        tvs_to_schedule.push_back(dc);
-      }
-
       // Set MemoryType
       dc->setMemoryType(MemoryType::Local);
       d_smem->setMemoryType(MemoryType::Shared);
 
       // Set LoadStoreOp
+      // TODO: extend support when mma is not cast to half
+      NVF_ERROR(
+          dc->dtype() == DataType::Half && !dc->definition()->isA<MmaOp>(),
+          "We support smem_epilogue on hopper only when the output of mma is cast to half")
       d_smem->definition()->as<LoadStoreOp>()->setOpType(
           LoadStoreOpType::StMatrix);
       d->definition()->as<LoadStoreOp>()->setOpType(
           LoadStoreOpType::CpAsyncBulkTensorTile);
 
-      // Block Schedule and Parallelize
-      blockTileTensors(tvs_to_schedule);
-      parallelizeBlocks(tvs_to_schedule);
+      // We apply the transformations that common to all tv from
+      // mma output (post cast op) to the fusion output.
+      // Once these common tranforms have been applied to d and propagated
+      // to dc (thus all othe tv till mma outs), we apply dc specific scheduling
+      // and propagate it to mma outs (post cast) and then apply smem and gmem
+      // specific schedules to d_smem and d respectively.
```

Comment:
But is this necessary because `scheduleMmaOutputAllocation` expects its argument to be scheduled like mma output already?

Reply:
https://github.com/NVIDIA/Fuser/blob/2ea03aecd88d679065db4b8cb6be73e541b81fca/csrc/scheduler/hopper_multi_matmul.h#L187-L190
This comment could be clearer (e.g. what is `is_mma_result` doing?) but it seems that this schedules the loop domain of the tv just like the mma output.
https://github.com/NVIDIA/Fuser/blob/2ea03aecd88d679065db4b8cb6be73e541b81fca/csrc/scheduler/mma_utils.h#L199-L202
This part is doing the swizzle required by the instruction. I think it schedules the inner two non-reduction dims by merge/reorder/split and parallelizes warpgroups via TIDx.

I get that one needs to come before the other one after diving in here. But I don't understand why a transform propagation is clearer than just writing:
```c++
      blockTileTensors({dc, d});
      parallelizeBlocks({dc, d});
      transformLikeMmaOutput(d, /*is_mma_result=*/false);
      transformLikeMmaOutput(dc, /*is_mma_result=*/false);
      mma_utils::MmaSwizzler::scheduleMmaOutputAllocation(dc...);
```

Reply:
I am fine with that code snippet. I was under the impression propagation is preferred over loop over tv to apply transforms. I'll make that change.

Reply:
I think in this case it makes it less clear. Since it is just a couple lines and we are always propagating just to one other tv, I would prefer to be explicit here. Please also place comments in this section.

Reply:
`is_mma_result` corresponds with the case when the input `tv` for `transformLikeMmaOutput` is the output `TensorView` of `MmaOp`. This tensorview a the reduction axis for the k-dimension at `-1` position, so the axis argument for `split`, `merge`, `reorder` operations need to be shift by `-1`. All the transformations are the same.

Maybe add this ^^^ to `transformLikeMmaOutput` comment.